mesh/mesh.hpp

```diff
@@ -765,7 +765,8 @@ class Mesh
    /** @brief Return the mesh geometric factors corresponding to the given
        integration rule. */
    const GeometricFactors* GetGeometricFactors(const IntegrationRule& ir,
-                                               const int flags);
+                                               const int flags,
+                                               DofToQuad::Mode = DofToQuad::FULL);
```

Comment:
I think the default here should be to use the faster option, `DofToQuad::TENSOR`. Internally, if not possible, the computation can fall back to `FULL`. Setting the parameter to `FULL` will tell the method to always use that choice for the computation.

Reply:
If internally we can know when we can use `DofToQuad::TENSOR`, why not removing the default parameter and having the following "default" implementation:
```c++
// A heuristic to know if we can use DofToQuad::TENSOR, something else might be more appropriate
static DofToQuad::Mode GetFESBasisMode(const FiniteElementSpace *fes)
{
   return dynamic_cast<const TensorBasisElement *>(fes->GetFE(0)) != nullptr ? DofToQuad::TENSOR : DofToQuad::FULL;
}

// The default implementation
const GeometricFactors* GetGeometricFactors(const IntegrationRule& ir, const int flags)
{
   DofToQuad::Mode mode = GetFESBasisMode(GetNodalFESpace());
   return GetGeometricFactors(ir, flags, mode);
}
```

Reply:
The new parameter allows you to force non-tensor product evaluation for tensor product elements -- e.g. to experiment, or for any other reason.

To be precise, however, `DofToQuad::Mode` is not quite right here -- it specifies the same mode for all elements and when you have mixed meshes, you may want to use different modes for different types of elements. For that case we'll probably need to introduce a more general version of this method -- we can do that when such need arises.

Reply:
> The new parameter allows you to force non-tensor product evaluation for tensor product elements -- e.g. to experiment, or for any other reason.

Yes, and I'm using it in the default implementation...

> To be precise, however, `DofToQuad::Mode` is not quite right here -- it specifies the same mode for all elements and when you have mixed meshes, you may want to use different modes for different types of elements. For that case we'll probably need to introduce a more general version of this method -- we can do that when such need arises.

Yes, this is exactly why I introduced a function that has this purpose. I sometimes have the impression that you have a design in mind for the future where we check for each element what we should do, can you clarify that this is *not* what you want to do.

Reply:
Okay, I now see that you still have version of `GetGeometricFactors` with the third `DofToQuad::Mode` parameter -- I didn't see that part earlier.

For mixed meshes (we want to support these in the future), I don't see how we can avoid checking every mesh element to see if it is tensor or not, if the desired `DofToQuad::Mode` is `TENSOR`. When it is `FULL`, no checks are needed. My suggestion is that we make `DofToQuad::Mode = TENSOR` the default, so we can get fast evaluations when possible by default. For now, the implementation can assume that all elements are of the same type and force an error if the mesh is mixed (this can be checked e.g. with `Mesh::GetNumGeometries`).

Reply:
I really think that we will need a multimap at some point, checking each element's type would restrict too much performance I think.

Reply:
`done`